File: app/src/main/java/org/michaelbel/moviemade/ui/modules/keywords/adapter/KeywordsAdapter.java
package org.michaelbel.moviemade.ui.modules.keywords.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.michaelbel.moviemade.R;
import org.michaelbel.moviemade.data.entity.Keyword;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.recyclerview.widget.RecyclerView;

public class KeywordsAdapter extends RecyclerView.Adapter<KeywordsAdapter.KeywordsViewHolder> {

    private List<Keyword> keywords = new ArrayList<>();

    public List<Keyword> getKeywords() {
        return keywords;
    }

    public void addKeywords(List<Keyword> results) {
        keywords.addAll(results);
        notifyItemRangeInserted(keywords.size() + 1, results.size());
    }

    @NonNull
    @Override
    public KeywordsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.chip_keyword, parent, false);
        KeywordsViewHolder holder = new KeywordsViewHolder(view);
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull KeywordsViewHolder holder, int position) {
        Keyword keyword = keywords.get(position);
        holder.keywordName.setText(keyword.getName());
    }

    @Override
    public int getItemCount() {
        return keywords.size();
    }

    class KeywordsViewHolder extends RecyclerView.ViewHolder {

        AppCompatTextView keywordName;

        private KeywordsViewHolder(View view) {
            super(view);
            keywordName = view.findViewById(R.id.keyword_name);
        }
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/features/keywords/KeywordsAdapter.kt
package org.michaelbel.moviemade.presentation.features.keywords

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.chip_keyword.*
import org.michaelbel.moviemade.R
import org.michaelbel.moviemade.core.entity.Keyword
import java.util.*

class KeywordsAdapter(
        private val listener: Listener
): RecyclerView.Adapter<KeywordsAdapter.KeywordsViewHolder>() {

    interface Listener {
        fun onKeywordClick(keyword: Keyword)
    }

    private val keywords = ArrayList<Keyword>()

    fun addKeywords(results: List<Keyword>) {
        keywords.addAll(results)
        notifyItemRangeInserted(keywords.size + 1, results.size)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): KeywordsViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.chip_keyword, parent, false)
        return KeywordsViewHolder(view)
    }

    override fun onBindViewHolder(holder: KeywordsViewHolder, position: Int) {
        holder.bind(keywords[position])
    }

    override fun getItemCount() = keywords.size

    inner class KeywordsViewHolder(override val containerView: View):
            RecyclerView.ViewHolder(containerView), View.OnClickListener, LayoutContainer {

        fun bind(keyword: Keyword) {
            keywordName.text = keyword.name
            containerView.setOnClickListener(this)
        }

        override fun onClick(v: View?) {
            if (adapterPosition != RecyclerView.NO_POSITION) {
                listener.onKeywordClick(keywords[adapterPosition])
            }
        }
    }
}